File: packages/runtime-python/src/keyma/runtime/testing.py
"""Test/reference utilities — port of ``@keyma/runtime-js`` ``testing.ts``.

:class:`InMemoryAdapter` is a fully in-memory :class:`KeymaDatabaseAdapter` used by
the runtime's own test suite and by adapter/plugin packages. It supports the full
Mongo-style ``where`` operator set, field/embedded/populate projections, and native
filtered counts. ``matches`` / ``matches_op`` are the standalone filter evaluators.

``brand_schema`` / ``brand_service`` attach the generated metadata statics
(``Class.schema`` / ``Class.service``) to a hand-written class — used by tests and
codegen fallback, where generated classes instead carry the statics directly.

Imported as ``from keyma.runtime.testing import InMemoryAdapter, matches, matches_op``.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from .types import SchemaMetadata, ServiceMetadata


class InMemoryAdapter:
    def __init__(self) -> None:
        self.stores: Dict[str, Dict[str, Dict[str, Any]]] = {}
        self._counter = 0

    def _store_for(self, schema: SchemaMetadata) -> Dict[str, Dict[str, Any]]:
        name = schema["name"]
        s = self.stores.get(name)
        if s is None:
            s = {}
            self.stores[name] = s
        return s

    async def ensure_schema(self, schema: SchemaMetadata) -> None:
        self._store_for(schema)

    async def create(
        self, schema: SchemaMetadata, data: Dict[str, Any], projection: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        store = self._store_for(schema)
        id_ = data.get("id")
        if id_ is None:
            self._counter += 1
            id_ = f'{schema["name"]}-{self._counter}'
        record = {**data, "id": id_}
        store[id_] = record
        return self._apply_projection(record, projection) if projection is not None else record

    async def read(
        self, schema: SchemaMetadata, where: Dict[str, Any], projection: Optional[Dict[str, Any]] = None
    ) -> Optional[Dict[str, Any]]:
        match = next((r for r in self._store_for(schema).values() if matches(r, where)), None)
        if match is None:
            return None
        return self._apply_projection(match, projection) if projection is not None else match

    async def list(self, schema: SchemaMetadata, query: Dict[str, Any]) -> List[Dict[str, Any]]:
        results = [r for r in self._store_for(schema).values() if matches(r, query.get("where") or {})]
        if query.get("skip") is not None:
            results = results[query["skip"]:]
        if query.get("limit") is not None:
            results = results[: query["limit"]]
        projection = query.get("projection")
        if projection is not None:
            results = [self._apply_projection(r, projection) for r in results]
        return results

    async def update(
        self,
        schema: SchemaMetadata,
        where: Dict[str, Any],
        data: Dict[str, Any],
        projection: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        store = self._store_for(schema)
        for id_, r in list(store.items()):
            if matches(r, where):
                updated = {**r, **data, "id": id_}
                store[id_] = updated
                return self._apply_projection(updated, projection) if projection is not None else updated
        # No existing record matched. When the filter targets a concrete id, upsert
        # under it (insert-or-merge); otherwise the update is undefined.
        if isinstance(where.get("id"), str):
            id_ = where["id"]
            updated = {**data, "id": id_}
            store[id_] = updated
            return self._apply_projection(updated, projection) if projection is not None else updated
        raise Exception(f"No record matches where {where!r}")

    async def delete(self, schema: SchemaMetadata, where: Dict[str, Any]) -> None:
        store = self._store_for(schema)
        for id_, r in list(store.items()):
            if matches(r, where):
                del store[id_]
                return

    async def count(self, schema: SchemaMetadata, where: Optional[Dict[str, Any]] = None) -> int:
        return sum(1 for r in self._store_for(schema).values() if matches(r, where or {}))

    def _apply_projection(self, record: Dict[str, Any], projection: Dict[str, Any]) -> Dict[str, Any]:
        if projection.get("fields") is None and projection.get("populate") is None:
            return record
        result: Dict[str, Any] = {}
        for key, spec in (projection.get("fields") or {}).items():
            if spec == 1:
                result[key] = record.get(key)
            else:
                value = record.get(key)
                result[key] = self._apply_embedded_spec(value, spec) if isinstance(value, dict) else None
        for field, node in (projection.get("populate") or {}).items():
            value = record.get(field)
            if not isinstance(value, str):
                result[field] = None
                continue
            referenced = self._store_for(node["schema"]).get(value)
            if referenced is None:
                result[field] = None
            elif node.get("projection") is not None:
                result[field] = self._apply_projection(referenced, node["projection"])
            else:
                result[field] = referenced
        return result

    def _apply_embedded_spec(self, value: Dict[str, Any], spec: Dict[str, Any]) -> Dict[str, Any]:
        result: Dict[str, Any] = {}
        for key, sub in spec.items():
            if sub == 1:
                result[key] = value.get(key)
            else:
                nested = value.get(key)
                result[key] = self._apply_embedded_spec(nested, sub) if isinstance(nested, dict) else None
        return result


def _js_strict_eq(a: Any, b: Any) -> bool:
    """Mirror JS ``===``: bool is not a number, and objects/lists compare by
    identity (never structurally)."""
    if isinstance(a, bool) != isinstance(b, bool):
        return False
    if isinstance(a, (dict, list)) or isinstance(b, (dict, list)):
        return a is b
    return a == b


def _same_value_zero(a: Any, b: Any) -> bool:
    """Mirror JS ``Array.prototype.includes`` (SameValueZero): no bool/int
    coercion, objects/lists by identity, ``NaN`` equals ``NaN``."""
    if isinstance(a, bool) != isinstance(b, bool):
        return False
    if isinstance(a, (dict, list)) or isinstance(b, (dict, list)):
        return a is b
    if isinstance(a, float) and isinstance(b, float) and a != a and b != b:
        return True
    return a == b


def matches(record: Dict[str, Any], where: Dict[str, Any]) -> bool:
    """Evaluate a Mongo-style ``where`` filter against a record."""
    for key, spec in where.items():
        if key == "$and":
            if not isinstance(spec, list):
                return False
            for sub in spec:
                if not matches(record, sub):
                    return False
            continue
        if key == "$or":
            if not isinstance(spec, list):
                return False
            if not any(matches(record, s) for s in spec):
                return False
            continue
        if key == "$nor":
            if not isinstance(spec, list):
                return False
            if any(matches(record, s) for s in spec):
                return False
            continue
        field_value = record.get(key)
        if isinstance(spec, dict):
            op_entries = list(spec.items())
            is_op_expr = len(op_entries) > 0 and all(k.startswith("$") for k, _ in op_entries)
            if is_op_expr:
                if not all(matches_op(field_value, op, arg) for op, arg in op_entries):
                    return False
                continue
        # JS uses strict `!==` here: object/array specs never match by structure,
        # and bool is distinct from int.
        if not _js_strict_eq(field_value, spec):
            return False
    return True


def matches_op(value: Any, op: str, arg: Any) -> bool:
    """Evaluate a single Mongo-style comparison operator (JS strict semantics)."""
    if op == "$eq":
        return _js_strict_eq(value, arg)
    if op == "$ne":
        return not _js_strict_eq(value, arg)
    if op == "$in":
        return isinstance(arg, list) and any(_same_value_zero(value, x) for x in arg)
    if op == "$nin":
        return isinstance(arg, list) and not any(_same_value_zero(value, x) for x in arg)
    # Ordered comparisons: JS coerces a missing field to NaN (every comparison
    # false → record excluded). Python would raise on None, so guard explicitly.
    if op == "$gt":
        return value is not None and arg is not None and value > arg
    if op == "$gte":
        return value is not None and arg is not None and value >= arg
    if op == "$lt":
        return value is not None and arg is not None and value < arg
    if op == "$lte":
        return value is not None and arg is not None and value <= arg
    return False


# ── Metadata branding (tests / codegen fallback) ──────────────────────────────
#
# Attach the generated metadata statics (``Class.schema`` / ``Class.service``) to a
# hand-written class. Generated classes carry these directly; test classes brand
# them on after the fact.


def brand_schema(cls: Any, schema: SchemaMetadata) -> Any:
    """Brand a plain class with :class:`SchemaMetadata` (as ``cls.schema``)."""
    cls.schema = schema
    return cls


def brand_service(cls: Any, service: ServiceMetadata) -> Any:
    """Brand a plain class with :class:`ServiceMetadata` (as ``cls.service``)."""
    cls.service = service
    return cls

File: packages/runtime/src/testing.ts
/**
 * Test/reference utilities for `@keyma/runtime/schema`.
 *
 * `InMemoryAdapter` is a fully in-memory {@link KeymaDatabaseAdapter} used by
 * the runtime's own test-suite and by adapter/plugin packages (e.g.
 * `@keyma/plugin-acl-js`). It is the single canonical implementation: it
 * supports the full Mongo-style `where` operator set, field/embedded/populate
 * projections, and native filtered counts.
 *
 * `brandSchema`/`brandService` attach the generated metadata statics
 * (`schema`/`service`) to a hand-written class — used by tests and codegen
 * fallback, where generated bundles instead carry the statics directly.
 *
 * Exposed via the package subpath `@keyma/runtime/schema/testing`.
 */
import type {
    KeymaDatabaseAdapter,
    ListQuery,
    AdapterProjection,
    AdapterFieldSpec,
} from "./adapter.js";
import type {
    SchemaMetadata,
    SchemaClass,
    ServiceMetadata,
    ServiceClass,
} from "./types.js";

export class InMemoryAdapter implements KeymaDatabaseAdapter {
    public stores = new Map<string, Map<string, Record<string, unknown>>>();
    private counter = 0;

    private storeFor(schema: SchemaMetadata): Map<string, Record<string, unknown>> {
        let s = this.stores.get(schema.name);
        if (s === undefined) {
            s = new Map();
            this.stores.set(schema.name, s);
        }
        return s;
    }

    async ensureSchema(schema: SchemaMetadata): Promise<void> {
        this.storeFor(schema);
    }

    async create(
        schema: SchemaMetadata,
        data: Record<string, unknown>,
        projection?: AdapterProjection,
    ): Promise<Record<string, unknown>> {
        const store = this.storeFor(schema);
        const id = (data["id"] as string | undefined) ?? `${schema.name}-${++this.counter}`;
        const record = { ...data, id };
        store.set(id, record);
        return projection !== undefined ? this.applyProjection(record, projection) : record;
    }

    async read(
        schema: SchemaMetadata,
        where: Record<string, unknown>,
        projection?: AdapterProjection,
    ): Promise<Record<string, unknown> | null> {
        const match = [...this.storeFor(schema).values()].find((r) => matches(r, where));
        if (match === undefined) return null;
        return projection !== undefined ? this.applyProjection(match, projection) : match;
    }

    async list(
        schema: SchemaMetadata,
        query: ListQuery,
    ): Promise<Record<string, unknown>[]> {
        let results = [...this.storeFor(schema).values()].filter((r) =>
            matches(r, query.where ?? {}),
        );
        if (query.skip !== undefined) results = results.slice(query.skip);
        if (query.limit !== undefined) results = results.slice(0, query.limit);
        if (query.projection !== undefined) {
            const proj = query.projection;
            results = results.map((r) => this.applyProjection(r, proj));
        }
        return results;
    }

    async update(
        schema: SchemaMetadata,
        where: Record<string, unknown>,
        data: Record<string, unknown>,
        projection?: AdapterProjection,
    ): Promise<Record<string, unknown>> {
        const store = this.storeFor(schema);
        for (const [id, r] of store.entries()) {
            if (matches(r, where)) {
                const updated = { ...r, ...data, id };
                store.set(id, updated);
                return projection !== undefined
                    ? this.applyProjection(updated, projection)
                    : updated;
            }
        }
        // No existing record matched. When the filter targets a concrete id,
        // upsert under it (insert-or-merge); otherwise the update is undefined.
        if (typeof where["id"] === "string") {
            const id = where["id"];
            const updated = { ...data, id };
            store.set(id, updated);
            return projection !== undefined
                ? this.applyProjection(updated, projection)
                : updated;
        }
        throw new Error(`No record matches where ${JSON.stringify(where)}`);
    }

    async delete(schema: SchemaMetadata, where: Record<string, unknown>): Promise<void> {
        const store = this.storeFor(schema);
        for (const [id, r] of store.entries()) {
            if (matches(r, where)) {
                store.delete(id);
                return;
            }
        }
    }

    async count(schema: SchemaMetadata, where?: Record<string, unknown>): Promise<number> {
        return [...this.storeFor(schema).values()].filter((r) => matches(r, where ?? {}))
            .length;
    }

    private applyProjection(
        record: Record<string, unknown>,
        projection: AdapterProjection,
    ): Record<string, unknown> {
        if (projection.fields === undefined && projection.populate === undefined) {
            return record;
        }
        const result: Record<string, unknown> = {};
        for (const [key, spec] of Object.entries(projection.fields ?? {})) {
            if (spec === 1) {
                result[key] = record[key];
            } else {
                const value = record[key];
                result[key] =
                    typeof value === "object" && value !== null
                        ? this.applyEmbeddedSpec(value as Record<string, unknown>, spec)
                        : null;
            }
        }
        for (const [field, node] of Object.entries(projection.populate ?? {})) {
            const value = record[field];
            if (typeof value !== "string") {
                result[field] = null;
                continue;
            }
            const referenced = this.storeFor(node.schema).get(value) ?? null;
            if (referenced === null) {
                result[field] = null;
            } else if (node.projection !== undefined) {
                result[field] = this.applyProjection(referenced, node.projection);
            } else {
                result[field] = referenced;
            }
        }
        return result;
    }

    private applyEmbeddedSpec(
        value: Record<string, unknown>,
        spec: { [key: string]: AdapterFieldSpec },
    ): Record<string, unknown> {
        const result: Record<string, unknown> = {};
        for (const [key, sub] of Object.entries(spec)) {
            if (sub === 1) {
                result[key] = value[key];
            } else {
                const nested = value[key];
                result[key] =
                    typeof nested === "object" && nested !== null
                        ? this.applyEmbeddedSpec(nested as Record<string, unknown>, sub)
                        : null;
            }
        }
        return result;
    }
}

/** Evaluate a Mongo-style `where` filter against a record. */
export function matches(
    record: Record<string, unknown>,
    where: Record<string, unknown>,
): boolean {
    for (const [key, spec] of Object.entries(where)) {
        if (key === "$and") {
            if (!Array.isArray(spec)) return false;
            for (const sub of spec) {
                if (!matches(record, sub as Record<string, unknown>)) return false;
            }
            continue;
        }
        if (key === "$or") {
            if (!Array.isArray(spec)) return false;
            const any = (spec as Record<string, unknown>[]).some((s) => matches(record, s));
            if (!any) return false;
            continue;
        }
        if (key === "$nor") {
            if (!Array.isArray(spec)) return false;
            const any = (spec as Record<string, unknown>[]).some((s) => matches(record, s));
            if (any) return false;
            continue;
        }
        const fieldValue = record[key];
        if (typeof spec === "object" && spec !== null && !Array.isArray(spec)) {
            const opEntries = Object.entries(spec as Record<string, unknown>);
            const isOpExpr =
                opEntries.length > 0 && opEntries.every(([k]) => k.startsWith("$"));
            if (isOpExpr) {
                for (const [op, arg] of opEntries) {
                    if (!matchesOp(fieldValue, op, arg)) return false;
                }
                continue;
            }
        }
        if (fieldValue !== spec) return false;
    }
    return true;
}

/** Evaluate a single Mongo-style comparison operator. */
export function matchesOp(value: unknown, op: string, arg: unknown): boolean {
    switch (op) {
        case "$eq":
            return value === arg;
        case "$ne":
            return value !== arg;
        case "$in":
            return Array.isArray(arg) && (arg as unknown[]).includes(value);
        case "$nin":
            return Array.isArray(arg) && !(arg as unknown[]).includes(value);
        case "$gt":
            return (value as number) > (arg as number);
        case "$gte":
            return (value as number) >= (arg as number);
        case "$lt":
            return (value as number) < (arg as number);
        case "$lte":
            return (value as number) <= (arg as number);
        default:
            return false;
    }
}

// ── Metadata branding (tests / codegen fallback) ─────────────────────────────
//
// Helpers that attach the generated metadata statics (`schema`/`service`) to a
// plain class. Generated bundles carry these statics directly; hand-written test
// classes brand them on after the fact.

/** Brand a plain class with SchemaMetadata at runtime (tests / codegen fallback). */
export function brandSchema<T>(
    cls: new (value?: Partial<T>) => T,
    schema: SchemaMetadata,
): SchemaClass<T> {
    Object.defineProperty(cls, "schema", {
        value: schema,
        enumerable: false,
        writable: false,
        configurable: false,
    });
    return cls as SchemaClass<T>;
}

/** Brand a plain class with ServiceMetadata at runtime (tests / codegen fallback). */
export function brandService<C extends Function>(
    cls: C,
    service: ServiceMetadata,
): C & ServiceClass {
    Object.defineProperty(cls, "service", {
        value: service,
        enumerable: false,
        writable: false,
        configurable: false,
    });
    return cls as C & ServiceClass;
}
